mlflow/langchain/_langchain_autolog.py

```diff
@@ -85,35 +89,139 @@ def _invoke(self, *args, **kwargs):
+
+    if func_name in ["invoke", "batch", "stream", "ainvoke", "abatch", "astream"]:
+        # `config` is the second positional argument of runnable APIs such as
+        # invoke, batch, stream, ainvoke, abatch, and astream
+        # https://github.com/langchain-ai/langchain/blob/7d444724d7582386de347fb928619c2243bd0e55/libs/core/langchain_core/runnables/base.py
+        if len(args) >= 2:
+            config = args[1]
+            config = _get_runnable_config_with_callback(config, mlflow_tracer)
+            return (args[0], config, *args[2:]), kwargs
+        else:
+            config = kwargs.get("config")
+            kwargs["config"] = _get_runnable_config_with_callback(config, mlflow_tracer)
+        return args, kwargs
+
+    elif func_name == "__call__":
+        # `callbacks` is the third positional argument of chain.__call__ function
+        # https://github.com/langchain-ai/langchain/blob/7d444724d7582386de347fb928619c2243bd0e55/libs/langchain/langchain/chains/base.py#L320
+        if len(args) >= 3:
+            callbacks = args[2] or []
+            callbacks = _inject_callback(callbacks, mlflow_tracer)
+            return (*args[:2], callbacks, *args[3:]), kwargs
+        else:
+            callbacks = kwargs.get("callbacks") or []
+            kwargs["callbacks"] = _inject_callback(callbacks, mlflow_tracer)
+            return args, kwargs
+
+    elif func_name == "get_relevant_documents":
+        # callbacks is only available as kwargs in get_relevant_documents function
+        # https://github.com/langchain-ai/langchain/blob/7d444724d7582386de347fb928619c2243bd0e55/libs/core/langchain_core/retrievers.py#L173
+        callbacks = kwargs.get("callbacks") or []
+        kwargs["callbacks"] = _inject_callback(callbacks, mlflow_tracer)
+        return args, kwargs
+
     else:
-        yield args, kwargs
+        _logger.warning(f"Unsupported function `{func_name}`. Skipping injecting MLflow callbacks.")
+        return args, kwargs
+
+
+def _get_runnable_config_with_callback(
+    original_config: Union[None, RunnableConfig, List[RunnableConfig]],
+    new_callback: BaseCallbackHandler,
+) -> Union[RunnableConfig, List[RunnableConfig]]:
+    """
+    Create a new RunnableConfig (or a list of them) with the new callback injected.
+
+    This function MUST return a new RunnableConfig instance, instead of mutating the original
+    config. This is because the original config may be shared across different calls and in-place
+    modification may cause unexpected behaviors, e.g. double injection.
+
+    Args:
+        original_config: the original RunnableConfig passed by the user
+        new_callback: a new callback to be injected
+    """
+    if original_config is None:
+        return RunnableConfig(callbacks=[new_callback])
+    elif isinstance(original_config, list):
+        return [_get_runnable_config_with_callback(c, new_callback) for c in original_config]
+    elif isinstance(original_config, dict):
```

Comment:
Can we use `RunnableConfig` instead of `dict`?

Reply:
`RunnableConfig` is `TypedDict` so it doesn't support `isinstance`: `TypeError: TypedDict does not support instance and class checks`. At runtime, it is just treated as dictionary type.

Reply:
Let's add a comment then? It's a bit confused if dev see this

Reply:
SG!